bundles/org.eclipse.core.runtime/src/org/eclipse/core/internal/runtime/LogServiceFactory.java

```diff
@@ -0,0 +1,34 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Christoph Läubrich and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.core.internal.runtime;
+
+import org.eclipse.core.runtime.ILog;
+import org.osgi.framework.*;
+
+/**
+ *
+ */
+public class LogServiceFactory implements ServiceFactory<ILog> {
+
+	@Override
+	public ILog getService(Bundle bundle, ServiceRegistration<ILog> registration) {
+		return InternalPlatform.getDefault().getLog(bundle);
```

Comment:
It is odd that we call the InternalPlatform.getLog here.  That will double map the ILog instance.  Once in InternalPlatform logs map and another in the framework registry where it needs to keep a mapping of all the using bundles->ILog instance for the ServiceFactory.  We should just create a new ILog instance here, probably refactoring InternalPlatform.getLog so that the code that creates the ILog instance is in a separate method this factory can call (e.g. `InternalPlatform.createLog`) that does not cache the ILog instance.

Reply:
By the way, even though the API of service factory tells one don't need to cache the instance we most probably still need some kind of cache, or the Platform.getLog() needs to perform an OSGi lookup each time, on behalf of the callers BundleContext, the problem is that the code only assumes a class or a Bundle, but this does not means the Bundle is started and has a bundlecontext!

Reply:
Although, having IternalPlatform use the service registry risks it overflowing the usecount of the service if a bundle continuously calls getLog over and over, so needs additional thought before doing that.

Reply:
> I'm not sure why you would introduce something new that double caches the ILog instance.

The OSGi does cache the already cached object, but is this really an issue?

> It doesn't seem that arduous to manage the `ILog` instances correctly within the ServiceFactory

It is already managed in the InternalPlatform and that seem to assume that there is a singelton per **bundle** and I don't want to have two instances just because one calls the static method and someone else uses service injection.

> Once that is working and in place then I would suggest InternalPlatform.getLog should use the passed in Bundle's BundleContext to then get the ILog instance from the service registry so that everything goes through the ServiceFactory.

As described above this won't work. I can call Platform.getLog(Some.class) and if Some.class is from an **resolved** bundle then there will be no BundleContext and thus no way to lookup/cache it by OSGi.

Reply:
> Although, having IternalPlatform use the service registry risks it overflowing the usecount of the service if a bundle continuously calls getLog over and over, so needs additional thought before doing that.

Yes, so IternalPlatform still needs a cache, so whatever we do there will 'doube caching' that's by design here and I think we can't reliable prevent this unless we get rid of `Platform.getLog(...)`.
Yeah I know... gazillion calls will, never happen, will make consumer mad, ... anyways there should be an alternative and even though it is not perfect is at least is not worse for just an additional map reference per bundle.

Reply:
Would it be possible to register the class `Log` as implementation of the `ILog` service using `scope = ServiceScope.BUNDLE`?
As you said, that would require to rework `InternalPlatform.getLog(Bundle)` to obtain the `ILog` service instance in behalf of the the passed bundle.

Reply:
If there would be no legacy code (e.g. InternalPlatform.getLog(Bundle)) yes, in all other cases: no, as described here as well, a bundle can only have such an ILog if it is started, while the legacy code here do not require this.